pkg/api/bzz.go

```diff
@@ -252,81 +398,84 @@ FETCH:
 	}
 
 	// serve requested path
-	s.serveManifestEntry(w, r, address, me.Reference(), !feedDereferenced)
+	s.serveManifestEntry(w, r, address, me, !feedDereferenced)
 }
 
-func (s *server) serveManifestEntry(w http.ResponseWriter, r *http.Request, address, manifestEntryAddress swarm.Address, etag bool) {
-	var (
-		logger = tracing.NewLoggerWithTraceID(r.Context(), s.logger)
-		ctx    = r.Context()
-		buf    = bytes.NewBuffer(nil)
-	)
-
-	// read file entry
-	j, _, err := joiner.New(ctx, s.storer, manifestEntryAddress)
-	if err != nil {
-		logger.Debugf("bzz download: joiner read file entry %s: %v", address, err)
-		logger.Errorf("bzz download: joiner read file entry %s", address)
-		jsonhttp.NotFound(w, nil)
-		return
+func (s *server) serveManifestEntry(
+	w http.ResponseWriter,
+	r *http.Request,
+	address swarm.Address,
+	manifestEntry manifest.Entry,
+	etag bool,
+) {
+
+	additionalHeaders := http.Header{}
+	mtdt := manifestEntry.Metadata()
+	if fname, ok := mtdt[manifestEntryMetadataFilenameKey]; ok {
+		additionalHeaders["Content-Disposition"] =
+			[]string{fmt.Sprintf("inline; filename=\"%s\"", fname)}
 	}
-
-	_, err = file.JoinReadAll(ctx, j, buf)
-	if err != nil {
-		logger.Debugf("bzz download: read file entry %s: %v", address, err)
-		logger.Errorf("bzz download: read file entry %s", address)
-		jsonhttp.NotFound(w, nil)
-		return
+	if mimeType, ok := mtdt[manifestEntryMetadataContentTypeKey]; ok {
+		additionalHeaders["Content-Type"] = []string{mimeType}
 	}
-	fe := &entry.Entry{}
-	err = fe.UnmarshalBinary(buf.Bytes())
-	if err != nil {
-		logger.Debugf("bzz download: unmarshal file entry %s: %v", address, err)
-		logger.Errorf("bzz download: unmarshal file entry %s", address)
-		jsonhttp.NotFound(w, nil)
-		return
+
+	s.downloadHandler(w, r, manifestEntry.Reference(), additionalHeaders, etag)
+}
+
+// downloadHandler contains common logic for dowloading Swarm file from API
+func (s *server) downloadHandler(w http.ResponseWriter, r *http.Request, reference swarm.Address, additionalHeaders http.Header, etag bool) {
+	logger := tracing.NewLoggerWithTraceID(r.Context(), s.logger)
+	targets := r.URL.Query().Get("targets")
+	if targets != "" {
+		r = r.WithContext(sctx.SetTargets(r.Context(), targets))
+
 	}
 
-	// read file metadata
-	j, _, err = joiner.New(ctx, s.storer, fe.Metadata())
+	reader, l, err := joiner.New(r.Context(), s.storer, reference)
 	if err != nil {
-		logger.Debugf("bzz download: joiner read file entry %s: %v", address, err)
-		logger.Errorf("bzz download: joiner read file entry %s", address)
+		if errors.Is(err, storage.ErrNotFound) {
+			logger.Debugf("api download: not found %s: %v", reference, err)
+			logger.Error("api download: not found")
+			jsonhttp.NotFound(w, nil)
+			return
+		}
+		logger.Debugf("api download: invalid root chunk %s: %v", reference, err)
```

Comment:
why? can it not be any other error?

Reply:
and why notFound error returned on 444?

Reply:
The error handling at L435 is special handling for "storage.ErrNotFound". But the second one is for the actual error at L434.
This is retained from old changes.

Reply:
I think it would be better to return a `500` in `L443`. I would also change the error from `invalid root chunk` to at least something like `unexpected error` or something like that. It's a corner case and we should at least be able to filter out 500s and try to figure out what caused this in the first place.